### PR TITLE
Playground UI improvements

### PR DESCRIPTION
**Pull request by BudsieBuds**
Hi everyone,

I've been lurking around here and seeing you all working on and building some awesome stuff. Just wanted to share what I've been working on for the last month or so. There's a lot unfinished and not working, I'm aware of it. I'm hoping for feedback and maybe some help.

The draft PR is a bit (very) messy. I've been using this branch as my playground while trying to create something that builds upon #3213. Even played around with updating the Markdown files, etc. If you're willing, please dig in and let me know what you think. What's working for you and where am I going in the wrong direction? Maybe some stuff is useable straight away in a smaller PR? All thoughts and questions are welcome.

So... what exactly did I try to do here:
- Conversion to Bootstrap 5 (#841)
- Responsive and more modern UI
- Conversion from Bower to npm (#1891)
- Added and (temporarily) removed some dependencies
   ...and updated all npm dependencies to the latest versions
- Kind of a lot of other stuff (and experiments)...

What's definitely broken here:
- Notification messages (toasts), should be converted to the new [Bootstrap included toasts](https://getbootstrap.com/docs/5.1/components/toasts/).
- Bootstrap-table is not yet working for some reason, this breaks a lot of screens.
- The clock.
- Adding or deleting config items, like rewards.
- A lot more, some of it is included in the TODO.md.

I've been looking for an easier way to share this with you guys, f.e. Docker Dev Environments. But everything I found has to be payed for, unfortunately.

### CODE REVIEW

**Comment by jekkos** [comment]
Thanks for this PR @BudsieBuds . I must say the diff is really huge. Some reformatting seems to have happened which makes it hard to spot what was changed.

I would go over this if I would have the time but currently I'm planning to go on leave for about a month and so I'm preparing for this now.

A first nice PR would be the markdown updates. If you want I can try to extract those change in a separate branch and make a PR for that? Or do you feel comfortable enough to branch this of first? Then maybe go step by step from there, like a PR for the bower changes, then maybe one that does some UI improvements.

I think the tables often break due to JS minification issues. You can try to append a `?debug=true` in your url to see if that makes it work again?

**Comment by BudsieBuds** [comment]
> Thanks for this PR @BudsieBuds . I must say the diff is really huge. Some reformatting seems to have happened which makes it hard to spot what was changed.

Thanks for your reply @jekkos. And I agree, it's too messy. I've been thinking a lot about this for the last couple of days. I'll sort stuff out, split it up into multiple PRs, and document the changes better.

> I think the tables often break due to JS minification issues. You can try to append a `?debug=true` in your url to see if that makes it work again?

I stripped the minification, and debug for that matter, for now, so I don't think that can be the issue. You probably haven't got the time to run this PR right now, but if you do, you'll probably see what's happening.

**Comment by daN4cat** [comment]
Thanks for the contribution!

**Comment by objecttothis** [comment]
@BudsieBuds Thanks for your work.  If you're willing, let's pull individual feature changes off of this and place them into individual PRs.  Right now I'm working on getting the codebase upgraded to CI 4.1.3 which will change a lot of syntax since it's all namespaced now and several CI functions have changed signatures.  That said, I don't think it negates work, but it probably  means that when we port over each change into a new PR, you'll need to work from the CI 4.1.3 version and do it by hand rather than mass copy/paste.

**Comment by jekkos** [comment]
As a first thing i would like to have a look at the markdown changes, these should be easy to review and merge

File: application/config/config.php [hunk 1]
```diff
@@ -359,7 +359,7 @@
 | https://codeigniter.com/user_guide/libraries/encryption.html
 |
 */
-$config['encryption_key'] = getenv('ENCRYPTION_KEY') ? getenv('ENCRYPTION_KEY') : '';
+$config['encryption_key'] = getenv('ENCRYPTION_KEY') ? getenv('ENCRYPTION_KEY') : 'c92f05bd67dc9a5897162e2dd48ba8cf9e3d74f52828f2631843d29227cc60ac';
```
Comment (daN4cat) [inline application/config/config.php:362]: Please remove this

**Comment by BudsieBuds** [comment]
@objecttothis Yeah, I noticed you working on an upgrade to CI4. Very nice! I'm working on splitting this all up and creating separate PRs. This was intended for some feedback. I'd have rather threw something up here that could be run locally within a few clicks, so feedback would be easier to give. But that's something I couldn't figure out how to do.

> As a first thing i would like to have a look at the markdown changes, these should be easy to review and merge

Agreed, I've been working on the text changes first. Adding more and reviewing the rest. That'll be the first separate PR.

**Comment by jekkos** [comment]
@BudsieBuds Is there something else specific we can potentially split or have a look at here?

**Comment by BudsieBuds** [comment]
> @BudsieBuds Is there something else specific we can potentially split or have a look at here?

@jekkos Not directly. I've been working on some stuff that's split from here, mainly the conversion to BS5. Haven't had a lot of time recently, hopefully I'll find the time soon to submit a decent PR. It's on my list. 👍🏻

**Comment by jekkos** [comment]
Ok cool once you have something we can discuss further and perhaps setup a call.

**Comment by objecttothis** [comment]
I'm now hoping to move to four days per week working on the CI4 upgrade. I mention that so that you keep it on your radar that we will need to make these changes CI4 compatible at some point. I'm almost at a place where I can begin testing the CI4 branch and hopefully from there we can get code reviews done. I don't know if it's easier to develop these changes before and then migrate them to CI4 or develop them in CI4 framework and merge them in after.

**Comment by jekkos** [comment]
great to hear @objecttothis, the CI upgrade seems to have been quite a job. I wonder whether you also had to touch the views alot in this case? I guess the BS5 upgrade would come down to reshuffling some html or the like in the view part and not touch models/controllers too much. Is that something you can confirm @BudsieBuds ?

Of course there are still some plugins like the modal that will need to be upgraded.

**Comment by objecttothis** [comment]
> great to hear @objecttothis, the CI upgrade seems to have been quite a job. I wonder whether you also had to touch the views alot in this case? I guess the BS5 upgrade would come down to reshuffling some html in the view part and not touch models/controllers too much. Is that something you can confirm @BudsieBuds ?
> 
> Of course there are still some plugins like the modal that will need to be upgraded. Both will probably touch a lot of files.

Unfortunately there are a large number of most ly syntactical changes in the views. Nothing that is rocket science but there are a lot of them. I've been working mainly on views in recent commits so you can take a look if you're curious.

**Comment by BudsieBuds** [comment]
@jekkos @objecttothis That's right, kind of forgot about that. I slowed down work on the BS5 upgrade because the CI4 upgrade is actively being worked on. My coding experience isn't enough to easily import changes and work from there, I guess I easily lose overview then. With bigger PRs I tend to reapply my new code to the new master instead of working in the changes on my branch.

> I'm now hoping to move to four days per week working on the CI4 upgrade. I mention that so that you keep it on your radar that we will need to make these changes CI4 compatible at some point. I'm almost at a place where I can begin testing the CI4 branch and hopefully from there we can get code reviews done. I don't know if it's easier to develop these changes before and then migrate them to CI4 or develop them in CI4 framework and merge them in after.

Big thumbs up for this. 👍🏻

**Comment by jekkos** [comment]
@objecttothis I wonder about the code review, not sure if possible but i guess you have encountered some bugs or did some small rework here and there?

Its going to be a huge diff for sure. Might be good to try to keep functional changes aside from the syntactic upgrade work, if feasible ofc.

**Comment by objecttothis** [comment]
> @objecttothis I wonder about the code review, not sure if possible but i guess you have encountered some bugs or did some small rework here and there?
> 
> Its going to be a huge diff for sure. Might be good to try to keep functional changes aside from the syntactic upgrade work, just to ease the review a bit.

For both @BudsieBuds and my work, likely much of the reviews will surround regression testing and just simply trying to break it and then use those broken parts to look in the code for clues as to where it's broken.

**Pull request closed without merging**